### ISSUE TITLE
LineDecoder is accidentally quadratic: iter_lines() seems to hang forever

### ISSUE THREAD

**Issue by gtedesco-r7** [encode/httpx#2422]
When calling `Response.iter_lines()`, things can seem to hang forever.

The problem is that `LineDecoder` is quadratic in it's string copying behaviour. If a 31MB chunk with 18,768 lines is passed in to `LineDecoder()` then it takes 1m45s to process it vs 0.1s for a simple `text.splitlines()`. This is readily reproduced by simply using `LineDecoder()` to decode a file, like `/usr/share/dict/words` (do not attempt, you may be waiting until the heat death of the universe).

It may be a bug somewhere else that `iter_text()` is returning chunks that are too big.

But either way, you probably want to reconsider the wisdom of slicing the string to chop the beginning off it vs. just keeping an index.

You can slice the string before returning from the decoder so it's only done once per chunk.

Anyhoo, thanks for the great work!

**Comment by gtedesco-r7** [comment]
Just a quick follow-on, even if the chunks are already line-decoded, processing the 31MB file takes 6 seconds!!

It would probably be a lot easier and a lot faster to just do something like:
```python
if buf.endswith('\n'):
    yield from buf.splitlines()  # with nothing leftover
else:
    it = buf.splitlines()
    prev = next(it)
    for line in it:
        yield prev
        prev = line
    self.remainder = prev
```

**Comment by gtedesco-r7** [comment]
Actually, it's easier than that since `splitlines()` already returns a list.

```python

    it = buf.splitlines()
    prev = next(it)
    for line in it:
        yield prev
        prev = line
    self.remainder = prev```

This implementation seems to work correctly and outperform in all cases.

In the case where chunks are all tiny (3 bytes) and lines are long (up to 6KB), there appears to be a slight slowdown, not sure why. But we're talking about 12s to 13.4s.

**Comment by giannitedesco** [comment]
Actually, the unit tests clued me in that `text.splitlines()` is no good because you want to keep the delimiters in. Switched to re, and that makes the performance regression with tiny chunks and long lines go away (13s down to 9s on my test data). :tada:

**Comment by giannitedesco** [comment]
Ah but okay, there's some whacky behaviours to do with CRLF and CR line-feeds, I will work on the PR as I get some time :joy:

**Comment by giannitedesco** [comment]
Okay, so I've fixed up the tests, and glad I did because the `splitlines()` implementation was initially buggy when CRLF sequence was split across chunks which is an issue that the [requests implementation](https://github.com/psf/requests/blob/main/requests/models.py#L853-L885) looks like it fails to deal with. I've fixed that now and as long as you're OK with the returned elements being stripped of newlines now. It looks as though all correct, and all other semantics preserved.
